src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs

```diff
@@ -3,311 +3,820 @@
+                using (var writer = new BinaryWriter(stream, TextEncoding, leaveOpen: true))
                 {
-                    item.SetMetadata(MetadataKeys.BuildAction, asset.BuildAction.ToString());
-                    item.SetMetadata(MetadataKeys.CopyToOutput, asset.CopyToOutput.ToString());
-                    item.SetMetadata(MetadataKeys.PPOutputPath, asset.PPOutputPath);
+                    writer.Write(ProjectPath);
+                    writer.Write(TargetFrameworkMoniker);
+                    writer.Write(DisableTransitiveProjectReferences);
+                    writer.Write(DisableFrameworkAssemblies);
+                    writer.Write(MarkPackageReferencesAsExternallyResolved);
+                    writer.Write(ProjectLanguage ?? "");
+                    writer.Write(RuntimeIdentifier ?? "");
+                }
+
+                stream.Position = 0;
 
-                    if (!string.IsNullOrEmpty(MetadataKeys.OutputPath))
+                using (var hash = CreateSettingsHash())
+                {
+                    return hash.ComputeHash(stream);
+                }
+            }
+        }
+
+        private sealed class CacheReader : IDisposable
+        {
+            private BinaryReader _reader;
+            private string[] _metadataStringTable;
+
+            public CacheReader(ResolvePackageAssets task)
+            {
+                byte[] settingsHash = task.HashSettings();
+                BinaryReader reader = null;
+
+                try
+                {
+                    if (File.GetLastWriteTimeUtc(task.ProjectAssetsCacheFile) >= File.GetLastWriteTimeUtc(task.ProjectAssetsFile))
                     {
-                        item.SetMetadata(MetadataKeys.OutputPath, asset.OutputPath);
+                        reader = OpenCacheFile(task.ProjectAssetsCacheFile, settingsHash);
                     }
+                }
+                catch (IOException)
+                {
+                }
 
-                    if (!string.IsNullOrEmpty(asset.CodeLanguage))
+                if (reader == null)
+                {
+                    using (var writer = new CacheWriter(task))
                     {
-                        item.SetMetadata(MetadataKeys.CodeLanguage, asset.CodeLanguage);
+                        writer.Write();
                     }
-                });
 
-            NativeLibraries = RaisePackageAssets(
-                runtimeTarget,
-                package => package.NativeLibraries);
+                    reader = OpenCacheFile(task.ProjectAssetsCacheFile, settingsHash);
+                }
+
+                _reader = reader;
+                ReadMetadataStringTable();
+            }
+
+            private static BinaryReader OpenCacheFile(string path, byte[] settingsHash)
+            {
+                var stream = File.Open(path, FileMode.Open, FileAccess.Read, FileShare.Read);
+                var reader = new BinaryReader(stream, TextEncoding, leaveOpen: false);
```

Comment:
Why isn't this just a using?

Reply:
Because we want to keep the reader open for the caller to own unless the header is out-of-date.

Reply:
👍